File: websocketserver/messaging.rb
# This module is used for nothing more than sending messages.
# It doesn't care about state. It just knows how to send particular messages. 
# Do you get the message ?
module Messaging
  DEBUG = true
  DELIMITER = "%"
  LOGIN = "L"
  LEVEL = "E"
  BANNED = "B"
  CHAT = "C"
  QUIT = "Q"
  DIE = "D"
  MOVE = "M"
  BULLET = "A"
  ACCEPT_LOGIN = "K"
  SERVER_MESSAGE = "S"
  NPC_MOVE = "N"
  def self.send_accept_login_message(game,ws,user)
    message = ACCEPT_LOGIN + DELIMITER + user.id
    ws.send message
    if DEBUG
      puts message
    end
  end
  def self.send_move_message(game,ws,user)
    message = MOVE + DELIMITER + user.id + DELIMITER + user.name + DELIMITER + user.dir.to_s + DELIMITER + user.x.to_s + DELIMITER + user.y.to_s + DELIMITER + user.dead.to_s
    ws.send message
    if DEBUG
      puts message
    end
  end

  def self.send_quit_message(game,ws,id)
    message = QUIT + DELIMITER + id
    ws.send message
    if DEBUG
      puts message
    end
  end

  def self.send_npc_move_message(game,ws,npc)
    message = NPC_MOVE + DELIMITER + npc.id + DELIMITER + npc.name + DELIMITER + npc.asset.to_s + DELIMITER + npc.dir.to_s + DELIMITER + npc.x.to_s + DELIMITER + npc.y.to_s + DELIMITER + npc.dead.to_s
    ws.send message
    if DEBUG
      puts message
    end
  end

  def self.send_specific_npc_move_message(game,ws,npc,x,y)
  message = NPC_MOVE + DELIMITER + npc.id + DELIMITER + npc.name + DELIMITER + npc.asset.to_s + DELIMITER + npc.dir.to_s + DELIMITER + x.to_s + DELIMITER + y.to_s + DELIMITER + npc.dead.to_s
    ws.send message
    if DEBUG
      puts message
    end
  end

  def self.send_specific_move_message(game,ws,user,x,y)
    message = MOVE + DELIMITER + user.id + DELIMITER + user.name + DELIMITER + user.dir.to_s + DELIMITER + x.to_s + DELIMITER + y.to_s + DELIMITER + user.dead.to_s
    ws.send message
    if DEBUG
      puts message
    end
  end

  def self.send_server_message_message(game,ws,message)
    message = SERVER_MESSAGE + DELIMITER + message
    ws.send message
    if DEBUG
      puts message
    end
  end

  def self.send_bullet_message(game,ws,bullet)
    message = BULLET + DELIMITER + bullet.id + DELIMITER + bullet.dir.to_s + DELIMITER + bullet.x.to_s + DELIMITER + bullet.y.to_s + DELIMITER + bullet.level.file
    ws.send message
    if DEBUG
      puts message
    end
  end

  def self.send_level_message(game,ws,file)
    message = LEVEL + DELIMITER + file
    ws.send message
    if DEBUG
      puts message
    end
  end

  def self.send_die_message(game,ws,user)
    message = DIE + DELIMITER + user.id
    ws.send message
    if DEBUG
      puts message
    end
  end

  def self.send_banned_message(game,ws)
    message = BANNED
    ws.send message
    if DEBUG
      puts message
    end
  end

  def self.send_chat_message(game,ws,sender,message)
    message = CHAT + DELIMITER + sender + DELIMITER + message.to_s
    ws.send message
    if DEBUG
      puts message
    end
  end
end

File: webserver/static/client/Messaging.js
// This class is used for nothing more than handling messages.
// It doesn't care about state. It just knows what to do when it gets a message.
// Do you get the message ?
var Messaging = {};

Messaging.DELIMITER = "%";
Messaging.LOGIN = "L";
Messaging.LEVEL = "E";
Messaging.BANNED = "B";
Messaging.CHAT = "C";
Messaging.QUIT = "Q";
Messaging.DIE = "D";
Messaging.MOVE = "M";
Messaging.BULLET = "A";
Messaging.ACCEPT_LOGIN = "K";
Messaging.SERVER_MESSAGE = "S";
Messaging.NPC_MOVE = "N";

Messaging.handleAcceptLoginMessage = function(id) {
  selfId = String(id);
  startGame();
}

Messaging.handleMoveMessage = function(id,name,dir,x,y,dead) {
  x = parseInt(x);
  y = parseInt(y);
  dir = parseInt(dir);

  if (!muted) {
    soundMove.play();
  }
  if (doesEntityExist(id) === false) {
    p = createPlayer(id,name,dir,x,y);
    if (dead === "true") {
      p.die();
    }
  } else {
    player = getEntityById(id);
    if (x === -1 && y === -1) {
      //Server is telling us to destroy this entity.
      player.destroy();
    } else {
      player.move(dir,x,y);
    }
  }
}

Messaging.handleQuitMessage = function(id) {
  deleteEntity(id);
}

Messaging.handleNpcMoveMessage = function(id,name,asset,dir,x,y,dead) {
  x = parseInt(x);
  y = parseInt(y);
  dir = parseInt(dir);
  if (!muted) {
    soundMove.play();
  }
  if (doesEntityExist(id) === false) {
    npc = createNpc(id, name, ASSET_OLD_MAN, dir, x, y);
    if (dead === "true") {
      npc.die();
    }
  } else {
    npc = getEntityById(id);
    if (x === -1 && y === -1) {
      //Server is telling us to destroy this entity.
      npc.destroy();
    } else {
      npc.move(dir,x,y);
    }
  }
}

Messaging.handleServerMessageMessage = function(message) {
    triggerAdminMessage(message);
}

Messaging.handleBulletMessage = function(id,dir,x,y,file) {
  x = parseInt(x);
  y = parseInt(y);
  dir = parseInt(dir);
  
  // is this the right level for this bullet?
  if (file === levelPath) {
    if (!muted) {
      soundShoot.play();
    }
    if (doesEntityExist(id) === false) {
      createBullet(id,dir,x,y,file);
    } else {
      bullet = getEntityById(id);
      bullet.move(dir,x,y);
    }
  }
}

Messaging.handleLevelMessage = function(file) {
  loadLevel(file);
}

Messaging.handleDieMessage = function(id) {
  if (!muted) {
    soundDie.play();
  }
  killPlayer(id);
  if (id === getSelf().id) {
    dead = true;
  }
}

Messaging.handleBannedMessage = function() {
    $(".blackout").show();
    alert("You are banned.");
}

Messaging.handleChatMessage = function(sender,message) {
    player = getEntityById(sender);
    if (player===null) {
      return;
    }
    player.chat(message);
}